lib/Dialect/LTL/LTLFolds.cpp

```diff
@@ -0,0 +1,117 @@
+//===- LTLFolds.cpp -------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/LTL/LTLOps.h"
+#include "mlir/IR/PatternMatch.h"
+
+using namespace circt;
+using namespace ltl;
+using namespace mlir;
+
+/// Check if an attribute is an integer zero.
+static bool isConstantZero(Attribute attr) {
+  if (!attr)
+    return false;
+  if (auto intAttr = dyn_cast<IntegerAttr>(attr))
+    return intAttr.getValue().isZero();
+  return false;
+}
+
+//===----------------------------------------------------------------------===//
+// DelayOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult DelayOp::fold(FoldAdaptor adaptor) {
+  // delay(s, 0, 0) -> s
+  if (adaptor.getDelay() == 0 && adaptor.getLength() == 0)
+    return getInput();
+
+  return {};
+}
+
+LogicalResult DelayOp::canonicalize(DelayOp op, PatternRewriter &rewriter) {
+  // delay(delay(s, I), J) -> delay(s, I+J)
+  // delay(delay(s, I), J, N) -> delay(s, I+J)
+  // delay(delay(s, I, N), J) -> delay(s, I+J)
+  // delay(delay(s, I, N), J, M) -> delay(s, I+J, N+M)
+  if (auto innerOp = op.getInput().getDefiningOp<DelayOp>()) {
+    rewriter.updateRootInPlace(op, [&] {
+      op.getInputMutable().assign(innerOp.getInput());
+      op.setDelay(op.getDelay() + innerOp.getDelay());
+      if (op.getLength() && innerOp.getLength())
+        op.setLength(*op.getLength() + *innerOp.getLength());
+      else
+        op.removeLengthAttr();
+    });
+    return success();
+  }
+
+  // delay(concat(s, ...), N, M) -> concat(delay(s, N, M), ...)
+  if (auto concatOp = op.getInput().getDefiningOp<ConcatOp>()) {
+    SmallVector<Value> newConcatArgs(concatOp.getInputs());
+    newConcatArgs[0] =
+        rewriter.create<DelayOp>(concatOp.getLoc(), concatOp.getInputs()[0],
+                                 op.getDelayAttr(), op.getLengthAttr());
+    NamedAttrList newAttrs;
+    for (auto namedAttr : op->getAttrs())
+      if (!llvm::is_contained(op.getAttributeNames(), namedAttr.getName()))
+        newAttrs.push_back(namedAttr);
+    rewriter.replaceOpWithNewOp<ConcatOp>(op, newConcatArgs, newAttrs);
+  }
+
+  return failure();
+}
+
+//===----------------------------------------------------------------------===//
+// ConcatOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult ConcatOp::fold(FoldAdaptor adaptor) {
+  // concat(s) -> s
+  if (getInputs().size() == 1)
+    return getInputs()[0];
+
+  return {};
+}
+
+LogicalResult ConcatOp::canonicalize(ConcatOp op, PatternRewriter &rewriter) {
```

Comment:
Likewise.  Even if these are moved to tablegen, making them explicit named patterns makes debugging easier (canonicalizer has a name to print out when applying the pattern).